src/notte/controller/base.py

```diff
@@ -94,18 +99,43 @@ async def execute_interaction_action(self, action: InteractionAction) -> Browser
         if action.press_enter is not None:
             press_enter = action.press_enter
         # locate element (possibly in iframe)
-        locator = await locale_element(self.window.page, action.selector)
+        locator: Locator = await locale_element(self.window.page, action.selector)
         original_url = self.window.page.url
 
         action_timeout = self.window.config.wait.action_timeout
 
+        def is_code(text: str) -> bool:
```

Comment:
can you move that into a new `notte/utils/code.py` file ? Can you also add some unit tests ? I want to make sure that we are not detecting too much code. Unit tests should contain at least 3 full positive examples (i.e only code from 3 different languages). Then 3 full negative examples (i.e empty string, "hello word" and some long text, at least 3 paragraphs). Then at least 3 mixed example (natural text + code mixed together)

Reply:
Rationale: the controller should not be responsible of `detecting code`. The controller only contain code for `executing actions in playwright context`